### PR TITLE
Preallocate slices with a known capacity.

### PR DESCRIPTION
**Pull request by RobbieMcKinstry**
Enable the prealloc linter, which identifies slices with a known capacity, but are not preallocated, which results in unnecessary allocations and memcpys.

# Description

## Checklist

- [ ] N/A: enabling a linter. I have added tests that prove my fix is effective or that my feature works

- [ ] **N/A: not a user-facing change.** I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by RobbieMcKinstry** [comment]
bors try

File: sdk/go/common/resource/plugin/provider_plugin.go [hunk 1]
```diff
@@ -771,15 +771,15 @@ func (p *provider) Diff(urn resource.URN, id resource.ID,
...
-	var replaces []resource.PropertyKey
+	var replaces = make([]resource.PropertyKey, 0, len(resp.GetReplaces()))
 	for _, replace := range resp.GetReplaces() {
 		replaces = append(replaces, resource.PropertyKey(replace))
 	}
-	var stables []resource.PropertyKey
+	var stables = make([]resource.PropertyKey, len(resp.GetStables()))
 	for _, stable := range resp.GetStables() {
 		stables = append(stables, resource.PropertyKey(stable))
 	}
```
Comment (abhinav) [inline sdk/go/common/resource/plugin/provider_plugin.go:781]: this should have length zero

Comment (RobbieMcKinstry) [reply sdk/go/common/resource/plugin/provider_plugin.go:781]: Nice catch!

**Comment by Frassle** [comment]
This is such a Goism 😆 we couldn't possibly give you an efficient and generic `map` function, here run this extra linter and write the same code in a hundred call sites instead.

**Comment by RobbieMcKinstry** [comment]
Yeah, it's... not fun! :D

File: pkg/codegen/docs/gen.go [hunk 2]
```diff
@@ -896,7 +896,8 @@ func (mod *modContext) genConstructorPython(r *schema.Resource, argsOptional, ar
 		return getDockerImagePythonFormalParams()
 	}
 
-	var params []formalParam
+	// We perform at least three appends before iterating over input types.
+	params := make([]formalParam, 0, 3+len(mod.inputTypes))
```
Comment (abhinav) [inline pkg/codegen/docs/gen.go:900]:
```suggestion
	params := make([]formalParam, 0, 3+len(r.InputProperties))
```

iteration over mod.inputTypes is not the one doing the appends

**Review by abhinav — APPROVED** [review]
approved minus one fix

**Comment by RobbieMcKinstry** [comment]
If bors fails to merge, I'll rebase, FP, and try again.

**Comment by dixler** [comment]
bors merge

**Pull request merged**